### PR TITLE
set_data()

### PR DESCRIPTION
**Pull request by clewis7**
#190 

add public method to ImageWidget to change the data

### CODE REVIEW

File: fastplotlib/widgets/image.py [hunk 1]
```diff
@@ -855,6 +856,30 @@ def reset_vmin_vmax(self):
 
             self.vmin_vmax_sliders[i].set_state(state)
 
+    def set_data(self, new_data: Union[np.ndarray, List[np.ndarray]], reset_vmin_vmax: bool = True, reset_current_index: bool = True):
+        """Change data of widget"""
+        if reset_current_index:
```
Comment (kushalkolar) [inline fastplotlib/widgets/image.py:861]: change to `reset_indices`, which resets indices for all dimensions

File: fastplotlib/widgets/image.py [hunk 2]
```diff
@@ -855,6 +856,30 @@ def reset_vmin_vmax(self):
...
+
+        if isinstance(new_data, np.ndarray) and isinstance(self.plot, Plot):
+            if new_data.shape != self._data[0].shape:
+                raise ValueError(f"new data shape {new_data.shape} does not equal current data shape {self._data[0].shape}")
+            self._data[0] = new_data
+            self.plot.graphics[0].data = self._data[0][self.current_index["t"]]
+        else:
+            if len(self._data) != len(new_data):
+                raise ValueError(f"number of new data arrays {len(new_data)} must match current number of data arrays {len(self._data)}")
+            positions = list(product(range(self.plot.shape[0]), range(self.plot.shape[1])))
```
Comment (kushalkolar) [inline fastplotlib/widgets/image.py:873]: this is unecessary, `GridPlot` is iterable :smile:

File: fastplotlib/widgets/image.py [hunk 3]
```diff
@@ -855,6 +856,30 @@ def reset_vmin_vmax(self):
...
+            for i, (new_darray, darray) in enumerate(zip(new_data, self._data)):
```
Comment (kushalkolar) [inline fastplotlib/widgets/image.py:874]: just add the image widget's gridplot instance to the `zip`

File: fastplotlib/widgets/image.py (hunk 2 above)
Comment (kushalkolar) [reply fastplotlib/widgets/image.py:873]:
```python
for subplot in gridplot:
  subplot#
```

File: fastplotlib/widgets/image.py [hunk 4]
```diff
@@ -855,6 +856,30 @@ def reset_vmin_vmax(self):
...
+                if new_darray.shape != darray.shape:
+                    raise ValueError(f"new data shape {new_darray.shape} does not equal current data shape {darray.shape}")
+                self._data[i] = new_data[i]
+                self.plot[positions[i][0], positions[i][1]].graphics[0].data = self._data[i][self.current_index["t"]]
```
Comment (kushalkolar) [inline fastplotlib/widgets/image.py:878]: add a comment that we assume `graphics[0]` is always the "image widget manged data"

File: fastplotlib/widgets/image.py [hunk 5]
```diff
@@ -855,6 +856,30 @@ def reset_vmin_vmax(self):
...
+
+        if reset_vmin_vmax and len(self.vmin_vmax_sliders) != 0:
```
Comment (kushalkolar) [inline fastplotlib/widgets/image.py:880]: this should happen regardless of vmin vmax sliders existing, only based on the kwarg `reset_vmin_vmax`

**Review by kushalkolar — CHANGES_REQUESTED** [review]

**Comment by clewis7** [comment]
running into a small issue for setting the data when the ndims of the data increase beyond "txy"

setting the overall data:
`self._data[0] = new_data`

however, if the dims are higher than 3 and I am trying to update the current displayed image, I get broadcasting errors because the shapes become mismatched
`self.plot.graphics[0].data = self._data[0][self.current_index["t"]]`

would need to do something like 
`self.plot.graphics[0].data = self._data[0][self.current_index["t"]]self.current_index["z"]` etc.
for whatever dims you are capable of sliding through, but I'm not sure how to make this work if you have an indeterminate amount of dims you could be sliding through

File: fastplotlib/widgets/image.py (hunk 5 above)
Comment (kushalkolar) [reply fastplotlib/widgets/image.py:882]: Modify reset_vmin_vmax() so that it works when no sliders are available.

**Comment by kushalkolar** [comment]
Just do the 0th index for all dims upto the last 2. 

Also I just realized we shouldn't directly set the graphic data, if windows functions or something are being used etc. So I would just set the current_index property. But it doesn't do anything if the current index is the same! So you could make a private method that has the logic for the current index setter with a "force" argument to force the update.

**Comment by clewis7** [comment]
the examples I have to set_data are working locally, I think it is sufficient to just force a call to the current index setter in order to get the graphics to update

**Comment by kushalkolar** [comment]
Maybe it's true that a simple call to the current_index setter is sufficient to set the image graphic data, haven't seen that code in a while. If it works then it works! :D

**Comment by kushalkolar** [comment]
Maybe it's true that a simple call to the current_index setter is sufficient to set the image graphic data, haven't seen that code in a while. If it works then it works! :D

**Review by kushalkolar — APPROVED** [review]

**Pull request merged**